matrix-sdk-android/src/test/java/org/matrix/android/sdk/internal/session/pushers/DefaultAddPusherTaskTest.kt

```diff
@@ -0,0 +1,178 @@
+import org.matrix.android.sdk.internal.network.GlobalErrorReceiver
+import org.matrix.android.sdk.internal.util.awaitTransaction
+import java.net.SocketException
+
+private val A_JSON_PUSHER = JsonPusher(
+        pushKey = "push-key",
+        kind = "http",
+        appId = "m.email",
+        appDisplayName = "Element",
+        deviceDisplayName = null,
+        profileTag = "",
+        lang = "en-GB",
+        data = JsonPusherData(brand = "Element")
+)
+
+class DefaultAddPusherTaskTest {
+
+    private val pushersAPI = FakePushersAPI()
+    private val monarchy = FakeMonarchy()
+
+    private val addPusherTask = DefaultAddPusherTask(
+            pushersAPI = pushersAPI,
+            monarchy = monarchy.instance,
+            requestExecutor = FakeRequestExecutor(),
+            globalErrorReceiver = FakeGlobalErrorReceiver()
+    )
+
+    @Test
+    fun `given no persisted pusher when adding Pusher then updates api and inserts result with Registered state`() {
+        monarchy.givenWhereReturns<PusherEntity>(result = null)
+
+        runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+
+        pushersAPI.verifySetPusher(A_JSON_PUSHER)
+        monarchy.verifyInsertOrUpdate<PusherEntity> {
+            withArg { actual ->
+                actual.state shouldBeEqualTo PusherState.REGISTERED
+            }
+        }
+    }
+
+    @Test
+    fun `given a persisted pusher when adding Pusher then updates api and mutates persisted result with Registered state`() {
+        val realmResult = PusherEntity(appDisplayName = null)
+        monarchy.givenWhereReturns(result = realmResult)
+
+        runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+
+        pushersAPI.verifySetPusher(A_JSON_PUSHER)
+
+        realmResult.appDisplayName shouldBeEqualTo A_JSON_PUSHER.appDisplayName
+        realmResult.state shouldBeEqualTo PusherState.REGISTERED
+    }
+
+    @Test
+    fun `given a persisted push entity and SetPush API fails when adding Pusher then mutates persisted result with Failed registration state and rethrows error`() {
+        val realmResult = PusherEntity()
+        monarchy.givenWhereReturns(result = realmResult)
+        pushersAPI.givenSetPusherErrors(SocketException())
+
+        assertFailsWith<SocketException> {
+            runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+        }
+
+        realmResult.state shouldBeEqualTo PusherState.FAILED_TO_REGISTER
+    }
+
+    @Test
+    fun `given no persisted push entity and SetPush API fails when adding Pusher then rethrows error`() {
+        monarchy.givenWhereReturns<PusherEntity>(result = null)
+        pushersAPI.givenSetPusherErrors(SocketException())
+
+        assertFailsWith<SocketException> {
+            runBlocking { addPusherTask.execute(AddPusherTask.Params(A_JSON_PUSHER)) }
+        }
+    }
+}
+
+internal class FakePushersAPI : PushersAPI {
```

Comment:
this PR introduces the concept of fakes [(here's a nice refresher!)](https://blog.pragmatists.com/test-doubles-fakes-mocks-and-stubs-1a7491dfa3da), the main purpose in this context is to avoid hitting the real api and database and to provide some reusable assertion/verification helpers

I've left the implementations within the test file for easier discussion/reviewing, will extract out to their own package if the team is happy with them :crossed_fingers: